### PR TITLE
[functions]support current_timestamp

### PR DESCRIPTION
**Pull request by liukun4515**
# Which issue does this PR close?

Closes #.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/expr/src/built_in_function.rs [hunk 1]
```diff
@@ -687,6 +691,9 @@ impl BuiltinScalarFunction {
             }
             BuiltinScalarFunction::CurrentDate => Ok(Date32),
             BuiltinScalarFunction::CurrentTime => Ok(Time64(Nanosecond)),
+            BuiltinScalarFunction::CurrentTimestamp => {
+                Ok(Timestamp(Nanosecond, Some("+00:00".into())))
```
Comment (alamb) [inline datafusion/expr/src/built_in_function.rs:695]: Maybe we could combine this with the case for `Now` to show they are the same function (just aliased)

File: datafusion/expr/src/expr_fn.rs [hunk 2]
```diff
@@ -747,6 +747,7 @@ scalar_expr!(
 scalar_expr!(CurrentDate, current_date, ,"returns current UTC date as a [`DataType::Date32`] value");
 scalar_expr!(Now, now, ,"returns current timestamp in nanoseconds, using the same value for all instances of now() in same statement");
 scalar_expr!(CurrentTime, current_time, , "returns current UTC time as a [`DataType::Time64`] value");
+scalar_expr!(CurrentTimestamp, current_timestamp, ,"returns current timestamp in nanoseconds, using the same value for all instances of current_timestamp() in same statement");
```
Comment (alamb) [inline datafusion/expr/src/expr_fn.rs:750]: I recommend we also note this is an alias for `now()` so people don't get confused about when to use `now()` and when to use `current_timestamp`

**Review by alamb — COMMENTED** [review]
I think this PR needs some sort of test prior to merge (otherwise I think current_timestamp will be broken inadvertently)

since `current_timestamp()` seems to be an alias for `now()`` I recommend you either:
1. Implement a rewrite in the sql planner that rewrites `current_timestamp` to `now()` or
2. Update other places that `now()` is handled specially in the DataFusion codebase (like constant folding) so they also handle current_timestamp

Here are some places that use `now`:  https://github.com/search?q=repo%3Aapache%2Farrow-datafusion+now

Given my reading of how special now is, it might make sense to explore the rewrite route

**Comment by liukun4515** [comment]
> I think this PR needs some sort of test prior to merge (otherwise I think current_timestamp will be broken inadvertently)
> 
> since `current_timestamp()` seems to be an alias for `now()`` I recommend you either:
> 
> 1. Implement a rewrite in the sql planner that rewrites `current_timestamp` to `now()` or
> 2. Update other places that `now()` is handled specially in the DataFusion codebase (like constant folding) so they also handle current_timestamp
> 
> Here are some places that use `now`: https://github.com/search?q=repo%3Aapache%2Farrow-datafusion+now
> 
> Given my reading of how special now is, it might make sense to explore the rewrite route

Do you means we should rewrite the `current_timestamp` to `now()` in the logical phase? and delete the physical operation for the `current_timestamp`?

I thinks it's a good way to maintain the new function, we just provide the function in the logical phase, and the function will be converted to the now.

**Comment by alamb** [comment]
> Do you means we should rewrite the current_timestamp to now() in the logical phase? and delete the physical operation for the current_timestamp?

Yes, this is what I would recommend doing unless there is some reason we need a different physical representation.

**Comment by liukun4515** [comment]
I change this pr to the draft, if it is ok, will reopen.

**Comment by alamb** [comment]
FWIW we now have a pretty good story for aliases in scalar functions

Adding support for `current_timestamp` (to be an alias of `now()`) would be pretty straightforward

**Pull request closed without merging**